### DISCUSSION TITLE
Expressive syntax vs. view preprocessing: how to strike a balance?

### DISCUSSION THREAD

**Discussion by jgonggrijp** [mustache/spec#163] [Q&A]
Since the start of the spec's existence, this topic has come up in nearly every discussion. It touches on the soul of the Mustache language, or at least what various people think that soul should be. If we could form a consensus around some kind of explicit principle, I think that would help to guide future decisions.

We have a spectrum between two extremes, which I dub "purist" and "pragmatist". A purist prefers a minimalist grammar and preprocessing the data before feeding them to the template. A pragmatist prefers an expressive grammar and not needing to preprocess the data.

The killer argument for the purist side is that Mustache distinguishes itself from other template languages by having a minimalist and logic-free grammar. If users want bells and whistles, they have many other options to choose from. Mustache should not become like Handlebars, ERB or Jinja; that would defeat the point of the language.

The killer argument for the pragmatist side is that template authors are not always in control of the input data. @agentgt formulated it well in https://github.com/mustache/spec/pull/147#issuecomment-1793791169:

> (...) I believe there are largely two groups of users that are using Mustache:
> 
> * developers that can easily decorate/transform or modify the model
> * users that really cannot and have to work with the model that is provided
> 
> The latter group is the one that really suffers without having virtual keys, or lambdas. (...)

A radical purist might believe that lambdas should never have made it into the language. A radical pragmatist might believe that Mustache needs syntax for literal values, string operations and arithmetic expressions.

I would like to formulate the central question of this discussion as follows: **where on this spectrum should Mustache try to be?**

In our answers, let us appreciate that both killer arguments are true at the same time: minimalism is one of the main features that make Mustache unique *and* template authors are not always in control of the input data. We will be making a tradeoff, regardless of where we position ourselves on the spectrum.

**Comment by jgonggrijp** [comment]
Having a go at this myself, let me order some existing and hypothetical features from most purist at the top to most pragmatist at the bottom. (I did not put this in the opening post, because any such ordering is imprecise and subjective.)

1. interpolation tag
2. section tag
3. partial tag
4. interpolation tag with escaping disabled
5. dotted names
6. implicit iterator
7. lambdas
8. filters (#153)
9. inheritance
10. filters with access to fallback values ("power filters", (https://github.com/mustache/spec/discussions/154#discussioncomment-7482763)
11. power lambdas (#155) without stack access
12. dynamic names inside partial tags
13. skips using a quote suffix (`{{name'}}`, #154)
14. virtual keys (https://github.com/mustache/spec/pull/147#issuecomment-1793791169)
15. dynamic names inside section tags (`{{#*listName}}`)
16. power lambdas with stack access
17. key modifier to render a section once for a nonempty list (`{{#?list}}`, #147)
18. passing arguments to lambdas (`{{#i18n(greeting)}}How do you do?{{/i18n}}`)
19. `{{else}}`
20. dynamic names inside filter pipes (`{{date | *format | lowercase}}`)
21. literal values and expressions inside interpolation tags (`{{age + 2}}`)
22. dynamic names inside block tags (`{{$*customizedFragment}}`)

I would say that 1-9 are essential to not make Mustache harder to use than necessary, while 15-22 are unreasonable. I think the balance is between 12 and 13: dynamic partials are still acceptable, skip syntax is not.

As a guiding principle, I would propose that every new feature added to the language should *multiply* the possibilities but only *increment* the syntactic complexity and the cost of implementation. In other words, new features that reduce the need for preprocessing are acceptable in principle, but only if high orthogonality is preserved. Given multiple possible solutions for a problem, the most orthogonal option should be chosen.

Filters (8) do really well in this regard: they open a million possibilities that are currently only possible with preprocessing, but they require only one new grammar rule and it is just a function call on the implementation side.

The list nonemptiness key modifier (17) does very poorly: it solves only one problem, but still adds a rule to the grammar and a special case to the implementation.

Virtual keys (14) are only marginally better than that. A single change to the parser can recognize all virtual keys, but each virtual key (for example `@key`, `@index`, `@first`, `@root`, ...) requires separate implementation and solves only one problem.

**Comment by jgonggrijp** [comment]
Starting from https://github.com/mustache/spec/discussions/153#discussioncomment-7534215, a few comments went a little bit sideways and were rather about the current discussion (#163). Quoted here for reasier reference:

@agentgt 

> (...) I am hoping to migrate some Handlebars users back to Mustache that switched because of simple things like first/last.

> (...) I'm concerned with initial users, onboarding and configuration pain. Standardized doc and well thought out feature is easier to learn and more useful than specific tribal knowledge.
> 
> For example I didn't feel like diving into `Mustache.java` code base to get `-first` working even though @spullara showed it was possible to have auto enumeration. I just used JMustache that supports it OOB instead.
> 
> I'm concerned that new folks will just say fuck this can't even make a CSV w/o registering a filter ... to handlebars I go.
> 
> It is important in my mind to make Mustache applicable to wide audience or else it becomes irrelevant.
> 
> I know it can be a slippery slope but one or two bells (builtin feature) may go a long way on user retention and growth especially given I think it is the number one question on Stackoverflow for Mustache last time I checked.

@spullara 

> I would rather a logicless templating system exist for those that want it. Honestly if I wanted all this other stuff I would use handlebars or some other library with all the features put in the template rather than the backing code.

@agentgt 

> @spullara data bindings of first/last/index I propose would be an optional extension. I think filters would be as well. Mustache.java does not have to implement it.
> 
> Also logicless is not well defined. The language has lambdas. It has dynamic names. It has inheritance with blocks.
> 
> I would put it this way. I would trade lambdas, dynamic names, and inheritance for first/last because it is that common.
> 
> I would venture to say that at least 75% of our templates that are not just plain partials have some first/last in there. When its code generation that is not HTML it is 100%.
> 
> I want to make that exceptionally common case not error prone and not wasteful.
> 
> Anyway that is my take on it. I aim to please users and more of pragmatist I guess. If it isn't spec I'm ok with that but I would like to see a best practice.
> 
> I just don't want Mustache to get too ivory towery. I have been there with XSLT on that.

I have to say I lean more towards the position of @spullara in this case. I understand that having to register a filter may be considered a barrier by some users, but look at it from the other side: Mustache has been around for 14 years. All that time, there has been no standard solution for first/last, with many implementations having no solution whatsoever other than data preprocessing (mine included). Despite that, it is the most widely ported templating language and it has a large and faithful crowd of users. I think that proves that Mustache's notational minimalism is appreciated and that many users are willing to give up some bells and whistles in exchange for that. I know I do.

On top of that:

- Winning a market share race from Handlebars is not a meaningful goal, as far as I'm concerned. Maybe for some individual implementations, but certainly not for the spec.
- If you give users who prefer Handlebars what they want, Mustache will end up being a second Handlebars. Does that make the world any better? I think a world with two alternative solutions to choose from, is a much more varied and inspiring place than a world that has two nearly identical solutions.

**Reply by jgonggrijp** [reply]
@agentgt in https://github.com/mustache/spec/discussions/153#discussioncomment-7537199:

> As for porting most of the logic that does first and last out there with Mustache-like and regular template languages have the `first` and `last` as sibling keys and not children.
> 
> The mustache ones implementing first/last as companion / siblings and not wrappers with `value`:
> 
>     * Trimou - http://trimou.org/doc/latest.html#variables
> 
>     * JMustache - https://github.com/samskivert/jmustache#-first-and--last
> 
>     * JStachio - https://jstach.io/jstachio/#mustache_virtual_keys_iterable
> 
>     * Handlebars.java - https://handlebarsjs.com/api-reference/data-variables.html#data-variables

In my opinion, "everyone else does it that way" is not a reason to spec something. We can do our own thing and try our best to do it well. Finding the best way to do templates (and design in general) is also not a democratic process; sometimes you have to do something unconventional in order to achieve something greater.

Besides, you demonstrated that you would be able to implement the variant that you prefer (with `{{.}}` instead of `{{value}}`). I don't really see the problem.

**Reply by agentgt** [reply]
Yes but being contrarian for the sake of "purity", "logicless", "homogenous"  or whatever ill defined things can also be perilous and possibly hubris. Like are we really that much better at coming up things than all those other guys?

I agree though that at times it is good to be "last mover" and then try to make a better decision based on your ecosystem. Like I have said I'm from Java and that same little bit of we can wait and do it better and different I'm familiar with (e.g. dealing with `null` and string interpolation).

> Mustache has been around for 14 years. All that time, there has been no standard solution for first/last, with many implementations having no solution whatsoever other than data preprocessing (mine included). 

Yes and there are so many different solutions to the problem in so many implementations. I wasn't even hoping for it to be a standard but a best practice.

And users do matter. They matter to me. And what others do does matter. We do not live in a vacuum. I care what they are familiar with and I want to provide the safest option. Otherwise I'm just writing a library for no one but myself would use... for the sake of minimalism.

Speaking of minimalism and perhaps I'm in the minority on this. 

I like Mustache because it is syntactically minimal but I really like it because it is safe and can be made even safer (as shown with JStachio).

I don't consider current lambdas that create the templates on the fly safe. Filters I do a little bit more but I don't like the return "" if it fails. If someone mistypes `@first` they are going to find out pretty fast it is wrong. I'm not so sure with filters.

I like Mustache because it can be locked down. I feel data variables for first/last are easier to lockdown than filters but at least filters are safer than lambdas.

> Besides, you demonstrated that you would be able to implement the variant that you prefer (with {{.}} instead of {{value}}). I don't really see the problem.

I don't really have a problem with filters doing enumeration to handle the problem. I have a problem in that it is unlikely an agreed standard filter that does enumeration would be included. It is unlikely I can make it performant compared to the current virtual keys. It is unlikely I can make it decoupled from JStachio.

Also I just wanted to make sure we know all the limitations to using the enumeration filter approach so that we can express that to ... again users.

If we are not going to help users with their problems why even add things to the spec.

To make i18next (sorry to cherrypick) easier for one developers' desire? 

I just want to make sure we fix the most common problem easily.

**Reply by jgonggrijp** [reply]
> Yes but being contrarian for the sake of "purity", "logicless", "homogenous" or whatever ill defined things can also be perilous and possibly hubris. Like are we really that much better at coming up things than all those other guys?
> 
> I agree though that at times it is good to be "last mover" and then try to make a better decision based on your ecosystem. Like I have said I'm from Java and that same little bit of we can wait and do it better and different I'm familiar with (e.g. dealing with `null` and string interpolation).

Right. I did not mean to claim that we will certainly do a better job. Just saying that exploring may be worth it.

> Speaking of minimalism and perhaps I'm in the minority on this.

It is absolutely fine to have a minority opinion. The discussions are **not** meant to be democratic as far as I'm concerned. I even specifically disabled to default poll category for that reason. If you are the only one with a particular opinion, you should still be heard, because you might be able to convince others. Even if you don't, you might still inspire people in other ways.

> 
> I like Mustache because it is syntactically minimal but I really like it because it is safe and can be made even safer (as shown with JStachio).

Nice nuance. You're basically positioning safety as a third dimension besides expressiveness and minimalism.

(Not meaning that as a criticism. I make it explicit because I find it interesting.)

> (...) If someone mistypes `@first` they are going to find out pretty fast it is wrong.

How does JStachio currently signal that to the user?

> If we are not going to help users with their problems why even add things to the spec.

I feel this is a strawman. Helping users is done by writing good documentation. Documentation should be clear about what cannot be done and about how to do what can be done. Whatever ends up in the spec does not preclude good documentation.

> 
> To make i18next (sorry to cherrypick) easier for one developers' desire?

Nice punch under the belt there. 😉 But no, I already conceded that one in https://github.com/mustache/spec/discussions/155#discussioncomment-7515586.

> 
> I just want to make sure we fix the most common problem easily.

It is rather bold to claim that first/last detection is the most common problem. But for the sake of argument, let us agree that it is common.

I would say that whether the current element is the first or last element of the list, is a statement about that element (as well as about the list). Another way to put it, is that the first/last boolean is a derived value.

Hence, first/last detection is a subset of the more general problem of derived values. By extension, the derived values problem is more common than the first/last detection problem alone.

I honestly think that filters are a very easy solution to the derived values problem. What is a more straightforward way to obtain a derived value than calling a function?

The idea of using filters to derive values is not completely new or alien, either. There are other templating languages that do it as well. Jinja and Angular come to mind immediately, because I have used them.

I'm not saying your argument is wrong, but I disagree it necessarily leads to virtual keys. I just demonstrated it can be used to argue for filters as well.

**Reply by spullara** [reply]
Mustache is like JSON to me. Very simple with very little to learn. Of course there are reasons for something insanely complex like XML, YAML, etc. but that doesn't mean we need to reach feature parity with them.

Mustache, the JSON of template languages has a nice ring to it.

**Reply by jgonggrijp** [reply]
@spullara Thanks for that way of comparing Mustache to another domain. I find it inspiring.

I find JSON, while minimal, not nearly as pleasant to write compared to Mustache. I also don't think that a JSON of template languages would have partials, lambdas or a way to change delimiters. I agree that the syntactic simplicity is similar, but on the semantic level, Mustache is much more powerful. I feel that Mustache is more efficient in squeezing as much expressivity out of as little syntax as possible.

How about the Scheme of template languages? It has a simple, consistent, declarative and explicit syntax, too. It also has very powerful abstractions underlying that syntax. Mustache is perhaps not as expressive as Scheme yet, but I think that relatively minor tweaks to the syntax could lift it to that level (relative to the maximum attainable expressiveness for template languages, at least).

**Reply by agentgt** [reply]
@jgonggrijp I agree that Mustache is better characterized as Scheme. I like Schema a lot. I used it in 1999 and had some minor impacts on it. I also used XSLT which I would characterize as the XML of templating. I don't think there is a JSON analog and I don't think Mustache is equivalent.

I'm almost fully on board with filters. While I would like to keep Mustache as simple as possible and while I agree with @jgonggrijp that it has served people well for 15 years that doesn't mean the language can or should not grow.

I know I pushed virtual keys hard and I'm probably the "bells and whistle guy" now (btw I meant the cherry pick of i18next as playful teasing... at least you are not the bells and whistles guy 😄  ).

I did that because Scheme had a hard uptake because of missing features but I still think it was mostly successful.

Anyway I think filters are needed to keep the language growing whether that is by new users or new problems. I think it should have been the original lambda. And I meant what I said about living in a vacuum. The same challenges and problems 15 years ago are not the same as today. If we do not evolve we die. 

Guy Steele - the creator of Scheme:

https://youtu.be/lw6TaiXzHAE?t=762

BTW @jgonggrijp you will appreciate rewinding back a little on his thoughts on English language given our friend who is concerned with that.

**Reply by jgonggrijp** [reply]
> I know I pushed virtual keys hard and I'm probably the "bells and whistle guy" now (btw I meant the cherry pick of i18next as playful teasing... at least you are not the bells and whistles guy 😄 ).

No worries, I took no offense. Also, there is nothing wrong with being the bells and whistles guy!

> Guy Steele - the creator of Scheme:
> 
> https://youtu.be/lw6TaiXzHAE?t=762

"Growing a language". Awesome talk! Funny, inspiring and relevant. I recommend watching it to anyone who has not clicked that link yet.
